### PR TITLE
I2304 multiple models to solver

### LINKED ISSUE #2304: [Bug]: IDAKLU solver fails with CCCV example
### PyBaMM Version

develop (22.8)

### Python Version

3.9

### Describe the bug

Running DFN with CCCV experiment using IDAKLU solver leads to error at the end of the first cycle

### Steps to Reproduce

```python3
import pybamm
import matplotlib.pyplot as plt

pybamm.set_logging_level("NOTICE")
experiment = pybamm.Experiment(
    [
        (
            "Discharge at C/5 for 10 hours or until 3.3 V",
            "Rest for 1 hour",
            "Charge at 1 A until 4.1 V",
            "Hold at 4.1 V until 10 mA",
            "Rest for 1 hour",
        ),
    ]
    * 3
)
model = pybamm.lithium_ion.DFN()

sim = pybamm.Simulation(model, experiment=experiment, solver=pybamm.IDAKLUSolver())
sim.solve()
```

### Relevant log output

```shell
2022-09-15 15:40:34.360 - [NOTICE] callbacks.on_cycle_start(174): Cycle 1/3 (11.704 ms elapsed) --------------------
2022-09-15 15:40:34.360 - [NOTICE] callbacks.on_step_start(182): Cycle 1/3, step 1/5: Discharge at C/5 for 10 hours or until 3.3 V
2022-09-15 15:40:34.632 - [NOTICE] callbacks.on_step_start(182): Cycle 1/3, step 2/5: Rest for 1 hour
2022-09-15 15:40:34.764 - [NOTICE] callbacks.on_step_start(182): Cycle 1/3, step 3/5: Charge at 1 A until 4.1 V
2022-09-15 15:40:34.955 - [NOTICE] callbacks.on_step_start(182): Cycle 1/3, step 4/5: Hold at 4.1 V until 10 mA
2022-09-15 15:40:35.087 - [NOTICE] callbacks.on_step_start(182): Cycle 1/3, step 5/5: Rest for 1 hour

[IDAS ERROR]  IDACalcIC
  The linear solver setup failed unrecoverably.


[IDAS ERROR]  IDASolve
  At t = 1.18566, the linear solver setup failed unrecoverably.

2022-09-15 15:40:35.103 - [ERROR] callbacks.on_experiment_error(226): Simulation error: idaklu solver failed
```

### PR DESCRIPTION
**Pull request by martinjrobins**
# Description

IDAKLU solver cannot setup more than one model. We have decided to restrict the `BaseSolver` to only setup a single model at at time. If multiple models are passed in a `RuntimeError` is raised

Fixes #2304 

## Type of change

- [x] Bug fix (breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: CHANGELOG.md
Comment (tinosulzer) [inline CHANGELOG.md:20]: will need to go into new section now

File: pybamm/solvers/base_solver.py
Comment (tinosulzer) [inline pybamm/solvers/base_solver.py:787]: I think `self.models_set_up` (a dict) could now be `self._model_set_up` (a model) since only one model is allowed

**Review by tinosulzer — APPROVED** [review]
Thanks, looks good and sorry for slow review.
In longer-term future, it would be better if the model itself (rather than the solver) stored the setting up functions, since it seems a bit weird to restrict one model instance per solver. But I'm happy to merge this for now as it fixes the existing bug

**Comment by martinjrobins** [comment]
Thanks @tinosulzer! I'll make those fixes. FYI, I'm quite happy with one-model-one-solver approach. There is always going to be significant setup for any model-solver pairing, and it makes sense to have setup stuff done within an `__init__()` rather than anywhere else

File: CHANGELOG.md
Comment (martinjrobins) [reply CHANGELOG.md:20]: done

**Comment by tinosulzer** [comment]
Ok fair enough, if it isn't broken don't fix it I guess

Updates look good so happy to merge once tests pass (merging develop should do it)

**Pull request merged**
